From 267d330460b52077a45f4d64462f50e9e641612a Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Sun, 15 May 2022 16:56:17 +0300
Subject: [PATCH] [SW-86865] gaudi3: implement run and halt arc api

This patch implements run_halt_arc api which will be used to halt arc
and run arc as part of fit processing.

Change-Id: I9b8cf7cdb4108f1d07d6622563e87d65497303b2
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index e2eb7e0..1a0c658 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -23,6 +23,7 @@
 #include <common/pci_ids.h>
 #include <tpm2.h>
 #include <gaudi3_security.h>
+#include <hl_fit_if.h>
 
 #include <d0_psoc_global_conf_masks.h>
 
@@ -860,6 +861,37 @@
 	return false;
 }
 
+#define ARC_RUN_HALT_ACK_DELAY_US			100
+#define ARC_RUN_M							BIT(0)
+#define ARC_HALT_M							BIT(1)
+
+__unused static void run_halt_arc(int core_id, bool run)
+{
+	uint32_t run_halt_offset;
+
+	switch (core_id) {
+	case FIT_CORE_ID_D0_PSOC_ARC0:
+		run_halt_offset = mmD0_PARC_ARC0_CFG_RUN_HALT_REQ;
+		break;
+	case FIT_CORE_ID_D0_PSOC_ARC2:
+		run_halt_offset = mmD0_PARC_ARC2_CFG_RUN_HALT_REQ;
+		break;
+	default:
+		return;
+	}
+
+	/* Set ARC in run/halt state */
+	if (run)
+		regmap_writel(ARC_RUN_M, run_halt_offset);
+	else
+		regmap_writel(ARC_HALT_M, run_halt_offset);
+
+	udelay(ARC_RUN_HALT_ACK_DELAY_US);
+
+	/* Clear run/halt request */
+	regmap_writel(0x0, run_halt_offset);
+}
+
 void arcboot_start_image(uint64_t startaddr)
 {
 	void (*image_entry)(void);
